### PR TITLE
Fix LibraryVersions.txt

### LINKED ISSUE #2645: LibraryVersions.txt does not contain information about library versions
It only contains versions for cmake, python, and the compilers; SpECTRE versioning; some system information; and compiler flags.

```
CMake Version: 3.17.4
SpECTRE Version: 2020.12.7
SpECTRE branch: develop
SpECTRE hash: v2020.12.07
Hostname: ungoliant
Host system: Linux-5.4.60-gentoo
Host system version: 5.4.60-gentoo
Host system processor: x86_64
CMAKE_CXX_FLAGS: -fPIC -fdiagnostics-color=always -ferror-limit=1 -ftemplate-backtrace-limit=0 -march=native
CMAKE_CXX_LINK_FLAGS:  -L/usr/lib64 -fuse-ld=gold
CMAKE_CXX_FLAGS_DEBUG: -g -DSPECTRE_DEBUG
CMAKE_CXX_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_C_FLAGS: -fPIC -march=native
CMAKE_C_FLAGS_DEBUG: -g
CMAKE_C_FLAGS_RELEASE: -O3 -DNDEBUG
CMAKE_Fortran_FLAGS: -march=native
CMAKE_Fortran_FLAGS_DEBUG: -g
CMAKE_Fortran_FLAGS_RELEASE: -O3 -DNDEBUG -O3
CMAKE_EXE_LINKER_FLAGS: -march=native -no-pie
CMAKE_BUILD_TYPE: Debug
CMAKE_CXX_COMPILER: /usr/lib/llvm/10/bin/clang++-10
CMAKE_CXX_COMPILER_VERSION: 10.0.1
CMAKE_C_COMPILER: /usr/lib/llvm/10/bin/clang-10
CMAKE_C_COMPILER_VERSION: 10.0.1
CMAKE_Fortran_COMPILER: /usr/bin/gfortran
CMAKE_Fortran_COMPILER_VERSION: 7.5.0
Python Version: 3.7.9
```

### PR DESCRIPTION
**Pull request by fmahebert**
## Proposed changes

- Instead of appending each library's version and THEN writing the file anew (erasing all the contents), now write the file at the beginning of the build and then append each library's version
- Rename LibraryVersions.txt to BuildInfoAndVersions.txt to better reflect the content.

This closes #2645 with option (2) discussed in that issue.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: cmake/PrintUsefulCMakeInfo.cmake [hunk 1]
```diff
@@ -1,14 +1,40 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
 
+# First append useful info to BuildInfoAndLibraryVersions.txt
+file(APPEND
+  "${CMAKE_BINARY_DIR}/BuildInfoAndLibraryVersions.txt"
```
Comment (nilsvu) [inline cmake/PrintUsefulCMakeInfo.cmake]: Just `BuildInfo.txt`? That's easier to tell people when they are supposed to send you the file :) and version info is also build info.

File: cmake/SetupBuildInfoAndVersionFile.cmake [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+file(WRITE
+  "${CMAKE_BINARY_DIR}/BuildInfoAndLibraryVersions.txt"
+  "SpECTRE Version: ${SPECTRE_VERSION}\n"
+  "Git description: ${GIT_DESCRIPTION}\n"
+  "Git branch: ${GIT_BRANCH}\n"
+  "Git hash: ${GIT_HASH}\n"
```
Comment (nilsvu) [inline cmake/SetupBuildInfoAndVersionFile.cmake]: If you set up the file even earlier you can append the spectre version and git info in `SpectreLoadMetadata.cmake`. Up to you.

File: cmake/SetupCatch.cmake [hunk 3]
```diff
@@ -4,10 +4,10 @@
 find_package(Catch 2.8.0 REQUIRED)
 
 spectre_include_directories("${CATCH_INCLUDE_DIR}")
-message(STATUS "Catch include: ${CATCH_INCLUDE_DIR}")
-message(STATUS "Catch version: ${CATCH_VERSION}")
+message(STATUS "Catch incl: ${CATCH_INCLUDE_DIR}")
+message(STATUS "Catch vers: ${CATCH_VERSION}")
```
Comment (nilsvu) [inline cmake/SetupCatch.cmake]: Why abbreviate? Better write "version" I think. (Also for Boost)

**Review by nilsvu — CHANGES_REQUESTED** [review]
Thanks for cleaning this up :)

File: cmake/SetupCatch.cmake (hunk 3 above)
Comment (fmahebert) [reply cmake/SetupCatch.cmake]: For consistency with all our other SetupBlah cmake output 🙂... I would be just as happy to make everything output the unabbreviated text instead (vers/incl/libs -> version/include/library), but that's a bigger diff

Comment (nilsvu) [reply cmake/SetupCatch.cmake]: I think some of your diff in this PR uses `version`, so I suggest to just go with that. No need to refactor all occurrences in this PR.

Comment (fmahebert) [reply cmake/SetupCatch.cmake]: 🤷 okay, reverting this change

File: cmake/AddHostToBuildInfoFile.cmake [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+file(APPEND
```
Comment (nilsvu) [inline cmake/AddHostToBuildInfoFile.cmake]: Perhaps name this file SetSiteName or something like that and include the `site_name` command, or the other way round: inline the `file` command in the root CMakeLists.txt and delete this file?

File: cmake/ClearBuildInfoFile.cmake [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+# Create/clear file so contents can then be appended into it.
+file(WRITE "${CMAKE_BINARY_DIR}/BuildInfo.txt" "")
```
Comment (nilsvu) [inline cmake/ClearBuildInfoFile.cmake]: Just inline this command in the root CMakeLists.txt and delete this file? It's just one line.

**Review by nilsvu — DISMISSED** [review]
LGTM, please squash. Here are some more suggestions (completely optional).

Would you mind updating the file name `LibraryVersions` > `BuildInfo` in `.github/ISSUE_TEMPLATE.md` as well?

**Comment by fmahebert** [comment]
@nilsleiffischer Yes, good suggestions. On my branch (not yet pushed) I've also updated a lot of formaline-related code that used `LibraryVersions.txt` so it now uses `BuildInfo.txt`.

One question though: after renaming the CMake log file to `BuildInfo.txt`, there's a bit of a cognitive conflict with the unrelated `SetupBuildInfo.cmake`. We could rename the log file (again) BuildInfo -> CMakeInfo or BuildLog or similar... or alternatively rename SetupBuildInfo -> SetupInfoFromBuild (or better, I just can't think of a better name at the moment). Thoughts?

**Comment by nilsvu** [comment]
`SetupInfoFromBuild` sound ok to me. Or `SetupInformer`, referencing the library name?

**Comment by fmahebert** [comment]
Squashed in the latest changes. I like your suggestion of `SetupInformer.cmake` so I went with that.

I've reorganized the commits since the first time you looked... after the various changes I felt this new organization was cleaner.

File: CMakeLists.txt [hunk 6]
```diff
@@ -32,10 +35,8 @@ endif ()
 
 set(CMAKE_VERBOSE_MAKEFILE OFF)
 
-site_name(HOSTNAME)
-
-include(SpectreInitializeVariables)
 include(SpectreGetGitHash)
+include(SetSiteName)
```
Comment (nilsvu) [inline CMakeLists.txt]: This should probably be named `SpectreSetSiteName`

**Review by nilsvu — CHANGES_REQUESTED** [review]
LGTM, just this one thing.

Also, could you double-check that formaline stores the file content in the H5 file correctly? We test a formaline build on CI, but don't check the content of the build info. That would be a great addition to the CI job of course (invoke `--dump-build-info` and check the output is the same as the original BuildInfo.txt), but I'm not asking you do that in this PR.

**Comment by fmahebert** [comment]
Renamed to `SpectreSetSiteName`, squashed directly.

**Comment by fmahebert** [comment]
Test failure is a clang-tidy complaint from Formaline code that is unrelated to this PR... (I just renamed something in the file).

File: docs/Tutorials/Visualization.md [hunk 7]
```diff
@@ -101,8 +101,8 @@ cause the program to exit after the information has been dumped. You can use the
 `--help` flag to get more details on what flags are available. The flag
 `--dump-paths` will print to screen the environment variables `PATH`, `CPATH`,
 `LD_LIBRARY_PATH`, `LIBRARY_PATH`, and `CMAKE_PREFIX_PATH` at link time. The
-flag `--dump-library-versions` will print to screen the contents of the
-`LibraryVersions.txt` file generated by CMake. The flag `--dump-environment`
+flag `--dump-build-info` will print to screen the contents of the
```
Comment (nilsvu) [inline docs/Tutorials/Visualization.md]: There's one more `--dump-library-version` in line 99 above

**Review by nilsvu — CHANGES_REQUESTED** [review]
squash directly

File: docs/Tutorials/Visualization.md (hunk 7 above)
Comment (fmahebert) [reply docs/Tutorials/Visualization.md]: Oops, that one has a typo (no final 's') and so I missed it in the find-replace... thanks for noticing!

**Comment by fmahebert** [comment]
Squashed in the latest rename fix. Thanks for the detailed reviews @nilsleiffischer!

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**